rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/mod.rs

```diff
@@ -0,0 +1,132 @@
+            // A previous target doesn't exist if we're trying to go backward
+            // from sequence 0 or if nothing has been indexed yet.
+            Some(0) | None => None,
+            Some(s) => Some(TargetSnapshot {
+                sequence: s.saturating_sub(1),
+                at_block: self.at_block,
+            }),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct TargetSnapshot {
+    pub sequence: u32,
+    pub at_block: u32,
+}
+
+#[derive(Debug)]
+pub enum SyncDirection {
+    Forward,
+    Backward,
+}
+
+/// A cursor that prefers to sync forward, but will sync backward if there is nothing to
+/// sync forward.
+pub(crate) struct ForwardBackwardSequenceAwareSyncCursor<T> {
+    forward: ForwardSequenceAwareSyncCursor<T>,
+    backward: BackwardSequenceAwareSyncCursor<T>,
+    last_direction: SyncDirection,
+}
+
+impl<T: Sequenced + Debug> ForwardBackwardSequenceAwareSyncCursor<T> {
+    /// Construct a new contract sync helper.
+    pub async fn new(
+        latest_sequence_querier: Arc<dyn SequenceAwareIndexer<T>>,
+        db: Arc<dyn HyperlaneSequenceIndexerStore<T>>,
+        chunk_size: u32,
+        mode: IndexMode,
+    ) -> Result<Self> {
+        let (sequence_count, tip) = latest_sequence_querier
+            .latest_sequence_count_and_tip()
+            .await?;
+        let sequence_count = sequence_count.ok_or(ChainCommunicationError::from_other_str(
+            "Failed to query sequence",
+        ))?;
+        let forward_cursor = ForwardSequenceAwareSyncCursor::new(
+            chunk_size,
+            latest_sequence_querier.clone(),
+            db.clone(),
+            sequence_count,
+            tip,
+            mode,
+        );
+        let backward_cursor =
+            BackwardSequenceAwareSyncCursor::new(chunk_size, db, sequence_count, tip, mode);
+        Ok(Self {
+            forward: forward_cursor,
+            backward: backward_cursor,
+            last_direction: SyncDirection::Forward,
+        })
+    }
+}
+
+#[async_trait]
+impl<T: Sequenced + Debug> ContractSyncCursor<T> for ForwardBackwardSequenceAwareSyncCursor<T> {
+    async fn next_action(&mut self) -> Result<(CursorAction, Duration)> {
+        // TODO: Proper ETA for backwards sync
+        let eta = Duration::from_secs(0);
+        // Prioritize forward syncing over backward syncing.
+        if let Some(forward_range) = self.forward.get_next_range().await? {
+            self.last_direction = SyncDirection::Forward;
+            return Ok((CursorAction::Query(forward_range), eta));
+        }
+
+        if let Some(backward_range) = self.backward.get_next_range().await? {
+            self.last_direction = SyncDirection::Backward;
+            return Ok((CursorAction::Query(backward_range), eta));
+        }
+        // TODO: Define the sleep time from interval flag
+        return Ok((CursorAction::Sleep(Duration::from_secs(5)), eta));
+    }
+
+    fn latest_block(&self) -> u32 {
```

Comment:
the doc comment of this function is `The latest block that has been queried`, rather than the latest block of the chain. I feel like always returning the forward cursor's latest block violates that, and we should either change the doc comment or the name of the method to `latest_queried_block`

Reply:
agree this is confusing, changed to `latest_queried_block`. We can also spend more time on this in https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3265 - it's not clear to me that the latest queried block is the right proxy for healthy either